libs/dal/src/repositories/subscriber/subscriber.schema.ts

```diff
@@ -165,15 +164,31 @@ subscriberSchema.index({
  *    subscriberId: /on-boarding-subscriber/i,
  *  });
  */
-subscriberSchema.index({
-  subscriberId: 1,
-  _environmentId: 1,
-  _id: 1,
-});
+
+/*
+ * This index needs to be unique and exclude "_id" to prevent duplicate subscribers during concurrent creation attempts.
+ * This situation could occur if two attempts are made to create a subscriber with the same subscriberId (e.g., 2022) simultaneously.
+ * We want to ensure that the _id field is not included in the index to avoid scenarios where MongoDB's unique validation fails to prevent duplicates, such as:
+ * subscriberId_2022:environmentId_123:_id_123
+ * subscriberId_2022:environmentId_123:_id_1234
+ * We expect an exception to be thrown when attempting to create two subscribers with the same subscriberId (e.g., 2022) within the same environment.
+ */
+index(
+  {
+    subscriberId: 1,
+    _environmentId: 1,
+    deleted: 1,
+  },
+  { unique: true }
+);
```

Comment:
❓ **question:** Wouldn't this index be an issue in this case:
The client already has deleted subscriber `{ subscriberId: '123', _environmentId: '123', deleted: true }`. Then he will create a new one with the same values `{ subscriberId: '123', _environmentId: '123', deleted: false }`. And I think that the index will fail when he tries to delete that new subscriber because he already has one deleted.

Reply:
That's a valid observation! We should address this scenario through multiple approaches:

1. If one subscriber was already deleted, perform a hard delete on the old one, followed by a soft delete on the newly deleted subscriber.
2. revisit the necessity of soft deleting subscribers altogether.
3. Remove the 'deleted' status from the index.

In my opinion, it's reasonable to proceed with option 3. The number of deleted subscribers under the same subscriberId and environmentId should be relatively small, so the optimization gained from adding 'deleted' in the index might not be significant.

Reply:
Done